### PR TITLE
Add support for "opam switch -" (go to previous switch)

### PR DESCRIPTION
**Pull request by kit-ty-kate**
This implements https://github.com/ocaml/opam/issues/4688 with the same behaviour as `git switch -`

cc @CraigFe

### CODE REVIEW

File: src/format/opamFile.ml [hunk 1]
```diff
@@ -1297,6 +1299,7 @@ module ConfigSyntax = struct
     { t with installed_switches }
   let with_switch_opt switch t = { t with switch }
   let with_switch switch t = { t with switch = Some switch }
+  let with_previous_switch previous_switch t = { t with previous_switch }
```
Comment (rjbou) [inline src/format/opamFile.ml:1302]: According the convention, `with_x` takes `'x` value, while `with_x_opt` takes `'x option` value (like `with_switch`).

File: src/format/opamFile.ml [hunk 2]
```diff
@@ -1361,6 +1365,9 @@ module ConfigSyntax = struct
       if t.switch = None then with_switch sw t
       else Pp.bad_format "Multiple switch specifications"
     in
+    let with_previous_switch sw t =
+      with_previous_switch (Some sw) t
+    in
```
Comment (rjbou) [inline src/format/opamFile.ml:1370]: There is no specific behavior added, better raise it up, with `with_previous_switch`.

File: src/client/opamSwitchCommand.ml [hunk 3]
```diff
@@ -312,6 +312,11 @@ let create
   let update_config = update_config && not (OpamSwitch.is_external switch) in
   let comp_dir = OpamPath.Switch.root gt.root switch in
   let simulate = OpamStateConfig.(!r.dryrun) || OpamClientConfig.(!r.show) in
+  (* TODO: We might want to make some more tests (quotes?, ...) *)
+  let switch_str = OpamSwitch.to_string switch in
+  if String.equal switch_str "-" then
+    OpamConsole.error_and_exit `Bad_arguments
+      "The name of the switch is not valid";
```
Comment (rjbou) [inline src/client/opamSwitchCommand.ml:319]: This check need to be added in `OpamSwitch.check` function, that is called at each `OpamSwitch.of_string` (that is called by cli argument parsing, among other places.)

File: tests/reftests/switch-set.test [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+[default] synchronised from file://${BASEDIR}/REPO
+Now run 'opam upgrade' to apply any package updates.
+### opam switch create foo test
+
+<><> Installing new switch packages <><><><><><><><><><><><><><><><><><><><><><>
+Switch invariant: ["test"]
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> installed test.1
+Done.
```
Comment (rjbou) [inline tests/reftests/switch-set.test:17]: To simplify the test, you can create empty switches, no need of a package.

**Review by rjbou — CHANGES_REQUESTED** [review]
Thanks!
Some modifications suggested, and some doc missing (manpage, website).
Also, should we add an format upgrade function to convert already present switches named `-`, in case they exist? Or we can assume that they don't exist (it wasn't forbidden, so it can exist).

**Comment by kit-ty-kate** [comment]
All done. The switch name `-` now does not interfere with `opam switch -`. It is still possible to create such a switch and to set it (using either `opam switch set -` or `opam switch --cli=2.1 -`) so no issues converting.

I’m still trying to figure out the manpage/doc situation but other than that this should be ready

File: src/client/opamSwitchCommand.ml [hunk 5]
```diff
@@ -392,6 +392,13 @@ let switch lock gt switch =
     (OpamSwitch.to_string switch) (OpamSwitch.to_string switch)
     (OpamStd.Format.itemize OpamSwitch.to_string installed_switches)
 
+let switch_previous lock gt =
+  match OpamFile.Config.previous_switch gt.config with
+  | Some switch_name -> switch lock gt switch_name
```
Comment (rjbou) [inline src/client/opamSwitchCommand.ml:397]: should we check here previous switch existence, in order to display a more specific error message than `switch` one?

**Comment by rjbou** [comment]
On non avoiding `-` as a switch name, we loose correspondence between `opam switch x` and `opam switch set x`. It's not an issue itself, but we need it to be clear on switch doc (and later on release doc). Old behavior is well kept with the cli check.

File: src/client/opamSwitchCommand.ml (hunk 5 above)
Comment (kit-ty-kate) [reply src/client/opamSwitchCommand.ml:397]: I don’t think it’s necessary, the default error message from opam switch is perfectly fine i think

**Comment by kit-ty-kate** [comment]
> On non avoiding `-` as a switch name, we loose correspondence between `opam switch x` and `opam switch set x`. It's not an issue itself, but we need it to be clear on switch doc (and later on release doc). Old behavior is well kept with the cli check.

is the new manpage entry clear enough?

**Comment by rjbou** [comment]
It is clear, i wrote the comment before you last push :)

**Review by rjbou — APPROVED** [review]

File: src/client/opamCommands.ml [hunk 6]
```diff
@@ -2438,7 +2438,14 @@ let switch cli =
           shell_eval_invocation shell
             (opam_env_invocation ~switch:"SWITCH" ~set_opamswitch:true ())
             |> Manpage.escape));
-  ] @ mk_subdoc ~cli ~defaults:["","list";"SWITCH","set"] commands
+  ] @
+    mk_subdoc ~cli ~defaults:["","list";"SWITCH","set"]
+      ~extra_defaults:[
+        cli_from cli2_2, "-",
+        "Switches back to the previous switch (similar to $(b,git switch -)). \
```
Comment (kit-ty-kate) [inline src/client/opamCommands.ml:2445]:
```suggestion
        "Switches back to the previous non-local switch (similar to $(b,git switch -)). \
```

File: src/client/opamSwitchCommand.ml (hunk 5 above)
Comment (kit-ty-kate) [reply src/client/opamSwitchCommand.ml:397]:
```suggestion
  | Some switch_name ->
    if not (OpamGlobalConfig.switch_exists gt switch_name) then
      OpamConsole.error_and_exit `Not_found "The previous switch no longer exists";
    switch lock gt switch_name
```
Something like that?

File: src/client/opamSwitchCommand.ml [hunk 7]
```diff
@@ -392,6 +392,16 @@ let switch lock gt switch =
     (OpamSwitch.to_string switch) (OpamSwitch.to_string switch)
     (OpamStd.Format.itemize OpamSwitch.to_string installed_switches)
 
+let switch_previous lock gt =
+  match OpamFile.Config.previous_switch gt.config with
+  | Some switch_name ->
+    if not (OpamGlobalState.switch_exists gt switch_name) then
+      OpamConsole.error_and_exit `Not_found "The previous switch no longer exists";
```
Comment (rjbou) [inline src/client/opamSwitchCommand.ml:399]:
```suggestion
      OpamConsole.error_and_exit `Not_found "The previous switch %s no longer exists"
        (OpamSwitch.to_string switch);
```

**Review by rjbou — APPROVED** [review]

**Comment by rjbou** [comment]
We need to check for opam root, but otherwise, lgtm!

**Comment by rjbou** [comment]
I need to rework the opam root test first, to make it no opamroot bump specific

**Comment by rjbou** [comment]
Reworked in #4913 & #4926 for the bump

**Comment by AltGr** [comment]
> On non avoiding - as a switch name, we loose correspondence between opam switch x and opam switch set x. It's not an issue itself, but we need it to be clear on switch doc (and later on release doc). Old behavior is well kept with the cli check.

that correspondance was already not there (which is why `set` still exists), you could have switches called `create` or `set`.

**Comment by AltGr** [comment]
LGTM; I can suggest an idea to avoid having to bump the format though: order the `installed-switches:` field by order of most recently used first. This would be fully backwards-compatible and consistent, and could even maybe find other uses.

**Comment by AltGr** [comment]
Couldn't resist and tried an implementation :)
There are differences though, I cut the patch in 2 to make the test difference obvious in ac19c215

File: src/format/opamFile.ml [hunk 8]
```diff
@@ -1266,7 +1265,14 @@ module ConfigSyntax = struct
   let repositories t = t.repositories
   let installed_switches t = t.installed_switches
   let switch t = t.switch
-  let previous_switch t = t.previous_switch
+  let previous_switch t =
+    let not_current = match t.switch with
+      | Some s -> (<>) s
+      | None -> fun _ -> true
```
Comment (kit-ty-kate) [inline src/format/opamFile.ml:1271]:
```suggestion
      | Some switch -> fun x -> not (String.equal switch x)
      | None -> fun _ -> true
```

File: src/format/opamFile.ml [hunk 9]
```diff
@@ -1297,10 +1303,15 @@ module ConfigSyntax = struct
   let with_repositories repositories t = { t with repositories }
   let with_installed_switches installed_switches t =
     { t with installed_switches }
-  let with_switch_opt switch t = { t with switch }
-  let with_switch switch t = { t with switch = Some switch }
-  let with_previous_switch_opt previous_switch t = { t with previous_switch }
-  let with_previous_switch previous_switch t = { t with previous_switch = Some previous_switch }
+  let with_switch switch t =
+    let installed_switches =
+      switch :: List.filter ((<>) switch) t.installed_switches
```
Comment (kit-ty-kate) [inline src/format/opamFile.ml:1308]:
```suggestion
      switch :: List.filter (fun x -> not (String.equal switch x)) t.installed_switches
```

File: tests/reftests/switch-set.test [hunk 10]
```diff
@@ -70,10 +70,8 @@ Switch ${BASEDIR} and all its packages will be wiped. Are you sure? [Y/n] y
 ### opam switch remove foo --yes
 Switch foo and all its packages will be wiped. Are you sure? [Y/n] y
 ### opam switch -
-[ERROR] The previous switch 'foo' no longer exists
-# Return code 5 #
+Selecting opam switch -.
 ### opam switch - --show
-[ERROR] The previous switch 'foo' no longer exists
-# Return code 5 #
+Would select opam switch bar.
```
Comment (kit-ty-kate) [inline tests/reftests/switch-set.test:75]: I’m not a big fan of the new behaviour but given it’s only going to be used very rarely, I’m fine with it.

File: src/format/opamFile.ml [hunk 11]
```diff
@@ -1265,6 +1265,14 @@ module ConfigSyntax = struct
   let repositories t = t.repositories
   let installed_switches t = t.installed_switches
   let switch t = t.switch
+  let previous_switch t =
+    let not_current = match t.switch with
+      | Some switch -> fun x -> not (String.equal switch x)
```
Comment (kit-ty-kate) [inline src/format/opamFile.ml:1270]:
```suggestion
      | Some switch -> fun x -> not (OpamSwitch.equal switch x)
```

File: src/format/opamFile.ml [hunk 12]
```diff
@@ -1295,8 +1303,15 @@ module ConfigSyntax = struct
   let with_repositories repositories t = { t with repositories }
   let with_installed_switches installed_switches t =
     { t with installed_switches }
-  let with_switch_opt switch t = { t with switch }
-  let with_switch switch t = { t with switch = Some switch }
+  let with_switch switch t =
+    let installed_switches =
+      switch :: List.filter (fun x -> not (String.equal switch x)) t.installed_switches
```
Comment (kit-ty-kate) [inline src/format/opamFile.ml:1308]:
```suggestion
      switch :: List.filter (fun x -> not (OpamSwitch.equal switch x)) t.installed_switches
```

**Review by rjbou — APPROVED** [review]

**Pull request merged**